### PR TITLE
Fix git dependency

### LINKED ISSUE #20: Support git dependency for plugins
I created a lint_package and tried to use this package within my application:

### 1. Project structure
**lint_package structure**
pubspec.yaml
```
dependencies:
  analyzer: ^4.2.0
  analyzer_plugin: ^0.10.0
  custom_lint_builder: ^0.0.9+1
```
In my custom_lint file (`bin/custom_lint.dart`) I have several imports:
```
import 'dart:isolate';

import 'package:analyzer/dart/analysis/results.dart';
import 'package:analyzer/dart/ast/ast.dart';
import 'package:analyzer/dart/element/type.dart';
import 'package:analyzer/source/source_range.dart';
import 'package:analyzer_plugin/protocol/protocol_generated.dart';
import 'package:analyzer_plugin/utilities/change_builder/change_builder_core.dart';
import 'package:custom_lint_builder/custom_lint_builder.dart';
```

**application structure**
pubspec.yaml
```
dependencies:
  custom_lint:
  lint_package:
     git:
        url: [...]
        ref: [...]
```
analysis_options.yaml
```
analyzer:
  plugins:
    - custom_lint
 ```

### 2. Problem
When running `flutter pub get` or `dart run custom_lint` within my application I get the following error:
```
IsolateSpawnException: Unable to spawn isolate: Error: Error when reading '**/.pub_cache/**/lint_package/.dart_tool/package_config.json': No such file or directory
Error: Couldn't resolve the package 'analyzer' in 'package:analyzer/dart/analysis/results.dart'.
Error: Couldn't resolve the package 'analyzer' in 'package:analyzer/dart/ast/ast.dart'.
Error: Couldn't resolve the package 'analyzer' in 'package:analyzer/dart/element/type.dart'.
Error: Couldn't resolve the package 'analyzer' in 'package:analyzer/source/source_range.dart'.
Error: Couldn't resolve the package 'analyzer_plugin' in 'package:analyzer_plugin/protocol/protocol_generated.dart'.
Error: Couldn't resolve the package 'analyzer_plugin' in 'package:analyzer_plugin/utilities/change_builder/change_builder_core.dart'.
Error: Couldn't resolve the package 'custom_lint_builder' in 'package:custom_lint_builder/custom_lint_builder.dart'.
```

The `.dart_tool` directory is not included in my lint_package repository.

If I manually execute `flutter pub get` in my local copy of this package (`.pub-cache/**/lint_package`) everything works as expected.

### PR DESCRIPTION
**Pull request by TimWhiting**
Fix #20 

As long as the dependencies are in the dependant package's package_config.json we can resolve it using that. Which should be the case, unless the plugin is invalidly depending on a dev_dependency that it is using as a regular dependency.

Verified that it works by changing the example lint to have git dependencies on this [branch](https://github.com/TimWhiting/dart_custom_lint/tree/git_dependency_fix), which also has the hot-reload to make it easy to debug. (Cherry picked just the changes to solve this issue to this PR though).

Ready for review
@rrousselGit

### CODE REVIEW

File: packages/custom_lint/lib/src/analyzer_plugin/plugin_link.dart [hunk 1]
```diff
@@ -50,14 +50,21 @@ final _pluginLinkProvider = FutureProvider.autoDispose
     p.join(pluginRootPath, 'bin', 'custom_lint.dart'),
   );
 
+  var packageConfigPath =
+      p.join(pluginRootPath, '.dart_tool', 'package_config.json');
+  if (!File(packageConfigPath).existsSync()) {
+    // Fall back to the packageConfig of the package depending on the custom lint (fixes the git dependency problem)
+    // Requires that all the dependencies of the custom lint are at least dev_dependencies of the package depending on it,
+    // which should be the case unless the custom lint is using a dev_dependency in its code that should be a regular dependency
+    packageConfigPath = p.join('.dart_tool', 'package_config.json');
```
[the remark this answers is not in the document]
Comment (TimWhiting) [reply packages/custom_lint/lib/src/analyzer_plugin/plugin_link.dart]: At first it wasn't obvious to me how to do it, but I think I figured it out mostly in the latest commit.

File: packages/custom_lint/lib/src/analyzer_plugin/plugin_link.dart [hunk 2]
```diff
@@ -50,14 +50,21 @@ final _pluginLinkProvider = FutureProvider.autoDispose
     p.join(pluginRootPath, 'bin', 'custom_lint.dart'),
   );
 
+  var packageConfigPath =
+      p.join(pluginRootPath, '.dart_tool', 'package_config.json');
+  if (!File(packageConfigPath).existsSync()) {
```
[the remark this answers is not in the document]
Comment (TimWhiting) [reply packages/custom_lint/lib/src/analyzer_plugin/plugin_link.dart]: I still am using a fallback in the latest commit, but I think it should be unneccessary, not sure if I should just do a non-null assertion instead.

**Comment by TimWhiting** [comment]
Added an e2e test, and fixed some failing tests (failing before my changes), but not all tests pass yet. Depending on your time, I'm happy to let you take over this PR, since I'm only spending a few hours in the evening on open source projects.

**Comment by rrousselGit** [comment]
Thanks a lot. I'm happy to take over. I'm slowly catching up with all the work you've done on custom_lint

**Comment by rrousselGit** [comment]
Thinking about this solution more, I think there's a possible problem if multiple applications use the same plugin

They could have different `package_config` but use the same config, in which case the behavior could be unexpected.

**Comment by rrousselGit** [comment]
I'm fine with ignoring that problem for now. I'll merge this to fix the more immediate issue of custom_lint not working in release mode

**Comment by rrousselGit** [comment]
Thanks a lot for your work as always :)

File: packages/custom_lint/lib/src/analyzer_plugin/plugin_link.dart [hunk 3]
```diff
@@ -188,6 +212,12 @@ final pluginMetasForContextRootProvider = Provider.autoDispose
 
 // TODO extract magic value
       if (dependencyPubspec.hasDependency('custom_lint_builder')) {
+        ref
```
Comment (rrousselGit) [inline packages/custom_lint/lib/src/analyzer_plugin/plugin_link.dart:215]: This slipped my review, but that doesn't work. Providers aren't allowed to modify other providers during their init

**Pull request merged**